Use sendWithAsyncReply for WebFullScreenManagerProxy::BeganExitFullScreen

achristensen07 opened the pull request:
#### df5c88e585197caff17c1f372d08a82e5cc46e28
<pre>
Use sendWithAsyncReply for WebFullScreenManagerProxy::BeganExitFullScreen
<a href="https://rdar.apple.com/144813733">rdar://144813733</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=287664">https://bugs.webkit.org/show_bug.cgi?id=287664</a>

Reviewed by Eric Carlson.

WebFullScreenManager::DidExitFullScreen should only be called as a reply to BeganExitFullScreen,
so I replaced it with a completion handler.  There are a few cases where we used to proactively
send this reply to a question that had not yet been asked by the web content process to the UI
process, but now we wait for the question to be asked before sending the reply.  The correctness
of such cases is tested by unit tests like WKWebViewCloseAllMediaPresentations.ElementFullscreen.

This is a step towards getting fullscreen working properly with site isolation.

I also may have fixed the PlayStation fullscreen code by changing the default from failing to enter
to succeeding to enter immediately.  If they want an async task to happen during the entering
time, they can add a listener object.

* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(RunJavaScriptConfirmResultListener::create):
(RunJavaScriptConfirmResultListener::~RunJavaScriptConfirmResultListener):
(RunJavaScriptConfirmResultListener::call):
(RunJavaScriptConfirmResultListener::RunJavaScriptConfirmResultListener):
(WKPageSetFullScreenClientForTesting):
(WKPageDidExitFullScreen): Deleted.
(WebKit::RunJavaScriptConfirmResultListener::create): Deleted.
(WebKit::RunJavaScriptConfirmResultListener::~RunJavaScriptConfirmResultListener): Deleted.
(WebKit::RunJavaScriptConfirmResultListener::call): Deleted.
(WebKit::RunJavaScriptConfirmResultListener::RunJavaScriptConfirmResultListener): Deleted.
* Source/WebKit/UIProcess/API/C/WKPage.h:
* Source/WebKit/UIProcess/API/C/WKPageFullScreenClient.h:
* Source/WebKit/UIProcess/API/C/playstation/WKView.cpp:
(WKViewDidExitFullScreen):
(WKViewSetViewClient):
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
(WebKit::PageClientImpl::beganExitFullScreen):
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.h:
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseDidExitFullScreen):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::enterFullScreen):
(WebKit::PageClientImpl::beganExitFullScreen):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/WPEWebViewLegacy.cpp:
(WKWPE::ViewLegacy::ViewLegacy):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
(WKWPE::ViewPlatform::didExitFullScreen):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::beganExitFullScreen):
(WebKit::WebFullScreenManagerProxy::didExitFullScreen): Deleted.
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::beganExitFullScreen):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController exitFullScreen]):
(-[WKFullScreenWindowController beganExitFullScreenWithInitialFrame:finalFrame:completionHandler:]):
(-[WKFullScreenWindowController _completedExitFullScreen:]):
(-[WKFullScreenWindowController _exitFullscreenImmediately]):
(-[WKFullScreenWindowController _startToDismissFullscreenChanged:]):
(-[WKFullScreenWindowController _dismissFullscreenViewController:]):
(-[WKFullScreenWindowController beganExitFullScreenWithInitialFrame:finalFrame:]): Deleted.
(-[WKFullScreenWindowController _completedExitFullScreen]): Deleted.
(-[WKFullScreenWindowController _dismissFullscreenViewController]): Deleted.
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::beganExitFullScreen):
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.h:
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(-[WKFullScreenWindowController finishedEnterFullScreenAnimation:]):
(-[WKFullScreenWindowController beganExitFullScreenWithInitialFrame:finalFrame:completionHandler:]):
(-[WKFullScreenWindowController finishedExitFullScreenAnimationAndExitImmediately:]):
(-[WKFullScreenWindowController beganExitFullScreenWithInitialFrame:finalFrame:]): Deleted.
* Source/WebKit/UIProcess/playstation/PageClientImpl.cpp:
(WebKit::PageClientImpl::beganExitFullScreen):
* Source/WebKit/UIProcess/playstation/PageClientImpl.h:
* Source/WebKit/UIProcess/playstation/PlayStationWebView.cpp:
(WebKit::PlayStationWebView::willEnterFullScreen):
(WebKit::PlayStationWebView::beganExitFullScreen):
(WebKit::PlayStationWebView::didExitFullScreen): Deleted.
* Source/WebKit/UIProcess/playstation/PlayStationWebView.h:
* Source/WebKit/UIProcess/win/PageClientImpl.cpp:
(WebKit::PageClientImpl::beganExitFullScreen):
* Source/WebKit/UIProcess/win/PageClientImpl.h:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::exitFullScreenForElement):
(WebKit::WebFullScreenManager::willExitFullScreen):
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.messages.in:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::willEnterFullScreen):
(WTR::TestController::beganExitFullScreen):
(WTR::TestController::finishFullscreenExit):
(WTR::TestController::resetStateToConsistentValues):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):

Canonical link: <a href="https://commits.webkit.org/290536@main">https://commits.webkit.org/290536@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8bfe60be8aad1f704c26d337b76a2e395ae47b76

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90373 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9903 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45300 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95379 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41149 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92425 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10290 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18221 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/69572 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27150 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93374 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7885 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81989 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49929 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/7620 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36352 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40281 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77938 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37422 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97211 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17561 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/78576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17818 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77812 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77788 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22234 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20850 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10828 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14209 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17571 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17312 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20764 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19096 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebKit/UIProcess/API/C/playstation/WKView.cpp:
```diff
@@ -174,9 +167,7 @@ void WKViewSetViewClient(WKViewRef view, const WKViewClientBase* client)
-
-            // FIXME: Replace this and WKViewWillEnterFullScreen with a listener object.
-            completionHandler(false);
+            completionHandler(true);
```
Was the old code incorrect, or has the meaning of the completion handler parameter changed? In either case, it would be a good idea to mention this change in the change log for future code spelunkers.

achristensen07 replied:
I've been talking with the maintainers of this code.  This makes it work by default instead of not working by default.  Will mention in the change log.

eric-carlson commented on Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
```diff
@@ -316,10 +300,26 @@ void WebFullScreenManagerProxy::beganEnterFullScreen(const IntRect& initialFrame
+        RefPtr page = m_page.get();
+        if (!page)
+            return completionHandler();
+        ALWAYS_LOG(LOGIDENTIFIER);
```
Nit: it is better to pass a variable initialized with `LOGIDENTIFIER` into the completion handler because `__func__` inside of a completion handler doesn't generate a method name that makes sense for our logging.

achristensen07 committed:
Use sendWithAsyncReply for WebFullScreenManagerProxy::BeganExitFullScreen
rdar://144813733
https://bugs.webkit.org/show_bug.cgi?id=287664

Reviewed by Eric Carlson.

WebFullScreenManager::DidExitFullScreen should only be called as a reply to BeganExitFullScreen,
so I replaced it with a completion handler.  There are a few cases where we used to proactively
send this reply to a question that had not yet been asked by the web content process to the UI
process, but now we wait for the question to be asked before sending the reply.  The correctness
of such cases is tested by unit tests like WKWebViewCloseAllMediaPresentations.ElementFullscreen.

This is a step towards getting fullscreen working properly with site isolation.

I also may have fixed the PlayStation fullscreen code by changing the default from failing to enter
to succeeding to enter immediately.  If they want an async task to happen during the entering
time, they can add a listener object.

* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(RunJavaScriptConfirmResultListener::create):
(RunJavaScriptConfirmResultListener::~RunJavaScriptConfirmResultListener):
(RunJavaScriptConfirmResultListener::call):
(RunJavaScriptConfirmResultListener::RunJavaScriptConfirmResultListener):
(WKPageSetFullScreenClientForTesting):
(WKPageDidExitFullScreen): Deleted.
(WebKit::RunJavaScriptConfirmResultListener::create): Deleted.
(WebKit::RunJavaScriptConfirmResultListener::~RunJavaScriptConfirmResultListener): Deleted.
(WebKit::RunJavaScriptConfirmResultListener::call): Deleted.
(WebKit::RunJavaScriptConfirmResultListener::RunJavaScriptConfirmResultListener): Deleted.
* Source/WebKit/UIProcess/API/C/WKPage.h:
* Source/WebKit/UIProcess/API/C/WKPageFullScreenClient.h:
* Source/WebKit/UIProcess/API/C/playstation/WKView.cpp:
(WKViewDidExitFullScreen):
(WKViewSetViewClient):
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
(WebKit::PageClientImpl::beganExitFullScreen):
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.h:
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseDidExitFullScreen):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::enterFullScreen):
(WebKit::PageClientImpl::beganExitFullScreen):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/WPEWebViewLegacy.cpp:
(WKWPE::ViewLegacy::ViewLegacy):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
(WKWPE::ViewPlatform::didExitFullScreen):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::beganExitFullScreen):
(WebKit::WebFullScreenManagerProxy::didExitFullScreen): Deleted.
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::beganExitFullScreen):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController exitFullScreen]):
(-[WKFullScreenWindowController beganExitFullScreenWithInitialFrame:finalFrame:completionHandler:]):
(-[WKFullScreenWindowController _completedExitFullScreen:]):
(-[WKFullScreenWindowController _exitFullscreenImmediately]):
(-[WKFullScreenWindowController _startToDismissFullscreenChanged:]):
(-[WKFullScreenWindowController _dismissFullscreenViewController:]):
(-[WKFullScreenWindowController beganExitFullScreenWithInitialFrame:finalFrame:]): Deleted.
(-[WKFullScreenWindowController _completedExitFullScreen]): Deleted.
(-[WKFullScreenWindowController _dismissFullscreenViewController]): Deleted.
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::beganExitFullScreen):
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.h:
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(-[WKFullScreenWindowController finishedEnterFullScreenAnimation:]):
(-[WKFullScreenWindowController beganExitFullScreenWithInitialFrame:finalFrame:completionHandler:]):
(-[WKFullScreenWindowController finishedExitFullScreenAnimationAndExitImmediately:]):
(-[WKFullScreenWindowController beganExitFullScreenWithInitialFrame:finalFrame:]): Deleted.
* Source/WebKit/UIProcess/playstation/PageClientImpl.cpp:
(WebKit::PageClientImpl::beganExitFullScreen):
* Source/WebKit/UIProcess/playstation/PageClientImpl.h:
* Source/WebKit/UIProcess/playstation/PlayStationWebView.cpp:
(WebKit::PlayStationWebView::willEnterFullScreen):
(WebKit::PlayStationWebView::beganExitFullScreen):
(WebKit::PlayStationWebView::didExitFullScreen): Deleted.
* Source/WebKit/UIProcess/playstation/PlayStationWebView.h:
* Source/WebKit/UIProcess/win/PageClientImpl.cpp:
(WebKit::PageClientImpl::beganExitFullScreen):
* Source/WebKit/UIProcess/win/PageClientImpl.h:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::exitFullScreenForElement):
(WebKit::WebFullScreenManager::willExitFullScreen):
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.messages.in:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::willEnterFullScreen):
(WTR::TestController::beganExitFullScreen):
(WTR::TestController::finishFullscreenExit):
(WTR::TestController::resetStateToConsistentValues):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):

Canonical link: https://commits.webkit.org/290536@main
